### PR TITLE
v2 - Add support for AAD password, AAD default, and AAD service principal auth

### PR DESCRIPTION
**Pull request by zijchen**
Maps [authentication types in connection strings](https://docs.microsoft.com/en-us/sql/connect/ado-net/sql/azure-active-directory-authentication?view=sql-server-ver16) to [authentication types in tedious driver](http://tediousjs.github.io/tedious/api-connection.html)
- Active Directory Password - 'azure-active-directory-password' (#60)
- Active Directory Default - 'azure-active-directory-default' (#93)
- Active Directory Service Principal - 'azure-active-directory-service-principal-secret' (#78)

Unit tests added to verify parsing of the different auth types, and the downstream config passed to tedious is correct.
"Integration" tests are added in pr-check.yml that verifies running this action with different auth types.

### CODE REVIEW

File: src/SqlConnectionConfig.ts [hunk 1]
```diff
@@ -67,12 +68,138 @@ export default class SqlConnectionConfig {
...
+            }
+            case 'activedirectorypassword': {
+                this._connectionConfig['authentication'] = {
+                    "type": 'azure-active-directory-password',
+                    "options": {
+                      // User and password should have been parsed already  
+                      "userName": this._connectionConfig.user,
+                      "password": this._connectionConfig.password,
+                      "clientId": core.getInput('client-id') || undefined,
+                      "tenantId": core.getInput('tenant-id') || undefined
```
Comment (zijchen) [inline src/SqlConnectionConfig.ts:136]: @dzsquared do you know if client-id or tenant-id can be exposed to log? Or should we mask them as secrets? passwords and secrets are masked already

Comment (zijchen) [reply src/SqlConnectionConfig.ts]: I masked them just in case

Comment (dzsquared) [reply src/SqlConnectionConfig.ts]: Minimally they're identifying information, but could be aggregated with other information to facilitate attack so let's keep them masked.  thank you!

File: src/SqlConnectionConfig.ts [hunk 2]
```diff
@@ -67,12 +68,148 @@ export default class SqlConnectionConfig {
...
         }
 
-        if (!this._connectionConfig.user) {
-            throw new Error(`Invalid connection string. Please ensure 'User' or 'User ID' is provided in the connection string.`);
+        const auth = this._connectionConfig['authentication'];
+        switch (auth?.type) {
+            case undefined: {
+                // SQL password
+                if (!this._connectionConfig.user) {
+                    throw new Error(`Invalid connection string. Please ensure 'User' or 'User ID' is provided in the connection string.`);
```
Comment (kisantia) [inline src/SqlConnectionConfig.ts:99]: should these errors be localized?

Comment (zijchen) [reply src/SqlConnectionConfig.ts:99]: Right now nothing is localized, and we haven't really had users requesting this either.

File: .github/workflows/pr-check.yml [hunk 3]
```diff
@@ -17,8 +17,18 @@ jobs:
     strategy:
       matrix:
         os: [windows-latest, ubuntu-latest]
+        auth_type: [sql_login, aad_password, service_principal] # Unique for each parallel job run, part of TEST_DB name to ensure no collisions from parallel jobs
+        include:
+          # Includes the connection string Github secret name, effectively a switch statement on auth_type
+          - auth_type: sql_login
+            connection_string_secret: AZURE_SQL_CONNECTION_STRING_NO_DATABASE
```
Comment (zijchen) [inline .github/workflows/pr-check.yml:24]: I think in a future PR I'll rework this so each part of the connection string is its own secret and we build the connection string at runtime. Right now you'll just have to trust that I didn't specify the same connection string 3 times 😆

File: __tests__/SqlConnectionConfig.test.ts [hunk 4]
```diff
@@ -13,7 +16,10 @@ describe('SqlConnectionConfig tests', () => {
             [`Server=test1.database.windows.net;User Id=user;Password="abc;1""2""adf(012j^72''asj;')'=33";Initial catalog=testdb`, 'validates values beginning with double quotes and also contains escaped double quotes', `abc;1"2"adf(012j^72''asj;')'=33`],
             [`Server=test1.database.windows.net;User Id=user;Password='ab""c;1''2''"''adf("0""12j^72''asj;'')''=33';Initial catalog=testdb`, 'validates values beginning with single quotes and also contains escaped single quotes', `ab""c;1'2'"'adf("0""12j^72'asj;')'=33`],
             [`Server=test1.database.windows.net;User Id=user;Password=JustANormal123@#$password;Initial catalog=testdb`, 'validates values not beginning quotes and not containing quotes or semi-colon', `JustANormal123@#$password`],
-            [`User Id=user;Password=JustANormal123@#$password;Server=test1.database.windows.net;Initial catalog=testdb`, 'validates connection string without server', `JustANormal123@#$password`]
+            [`User Id=user;Password=JustANormal123@#$password;Server=test1.database.windows.net;Initial catalog=testdb`, 'validates connection string without server', `JustANormal123@#$password`],
```
Comment (Benjin) [inline __tests__/SqlConnectionConfig.test.ts:19]: `Password=JustANormal123@#$password` doesn't trigger credscan at all?  Huh, unexpected; I thought it needed to include `placeholder`.

Comment (Benjin) [reply __tests__/SqlConnectionConfig.test.ts:19]: Unless there's a reason to use that password specifically, might be good (preemptive?) to use `placeholder` anyway.

Comment (Benjin) [inline __tests__/SqlConnectionConfig.test.ts:19]: Maybe I'm misunderstanding what "validates connection string without server" means, but there's a server URL in that connection string; it's just in a different spot...

File: action.yml [hunk 5]
```diff
@@ -5,8 +5,14 @@ inputs:
     description: 'Name of the Azure SQL Server name, like Fabrikam.database.windows.net.'
     required: false
   connection-string:
-    description:  'The connection string, including authentication information, for the Azure SQL Server database.'
+    description: 'The connection string, including authentication information, for the Azure SQL Server database.'
     required: true
+  client-id:
+    description: 'For Active Directory connections, the application (client) ID from the registered Azure application'
```
Comment (Benjin) [inline action.yml:11]: Should this be "Azure Active Directory" connections?

File: __tests__/SqlConnectionConfig.test.ts [hunk 6]
```diff
@@ -29,30 +35,130 @@ describe('SqlConnectionConfig tests', () => {
...
-            [`Server=test1.database.windows.net;User Id=user;Password=NotANormal123@;#$password;Initial catalog=testdb`, 'validates values not enclosed in quotes and containing semi-colon'],
-            [`Server=test1.database.windows.net;Password=password;Initial catalog=testdb`, 'missing user id'],
-            [`Server=test1.database.windows.net;User Id=user;Initial catalog=testdb`, 'missing password'],
-            [`Server=test1.database.windows.net;User Id=user;Password=password;`, 'missing initial catalog']
+            [`Server=test1.database.windows.net;User Id=user;Password="ab'=abcdf''c;123;Initial catalog=testdb`, `Invalid connection string. A valid connection string is a series of keyword/value pairs separated by semi-colons. If there are any special characters like quotes or semi-colons in the keyword value, enclose the value within quotes. Refer this link for more info on conneciton string https://aka.ms/sqlconnectionstring`, 'validates values beginning with double quotes but not ending with double quotes'],
```
[the remark this answers is not in the document]
Comment (kisantia) [reply __tests__/SqlConnectionConfig.test.ts]: ah that was actually a different typo you fixed haha. The one I had meant was "conneciton"

**Review by kisantia — APPROVED** [review]

**Pull request merged**